Add option to reuse database connections

RobinPapke opened the pull request:
- add optional connection argument to use connections for fetching data
- add execution method to write to database (with commit)

---

1. Example: Fetch using the same database connection
```python
db = Database()

# use same connection for multiple queries
with db.connect()  as connection:
    db.fetch(query, connection=connection)
    db.execute(query, connection=connection)

# open new connection for the query
db.fetch(query)
```
 ---

Regarding thread-based concurrency middleware, it's worth mentioning that thread safety is __not__ guaranteed for sharing connections between threads by default:
- PEP 249: https://www.python.org/dev/peps/pep-0249/#threadsafety
- pymysql: https://github.com/PyMySQL/PyMySQL/blob/master/pymysql/__init__.py#L43
- vertica-python: https://github.com/vertica/vertica-python/blob/master/vertica_python/__init__.py#L66

twheys commented:
Any kind of code that needs to start and end before and after a block of work should be implemented to use the python `with` keyword, otherwise you make the code prone to errors (what happens if you forget to close the session?) 

Can you take a look at the pep and change the implementation so that it can be done like

```
db = Database()

# create a connection on the database instance
with db.use_session():
    # fetch will automatically use the established database connection
    db.fetch(query)
```

RobinPapke commented:
@twheys I fixed the issues you mentioned. Connections can now be used as arguments for methods with db access and reused that way. Combined with the keyword ```with```,  the connection will be closed automatically outside of the context. I updated the example in the description accordingly.
